Pass data.path as upload source in storage

DawidJanczak opened the pull request:
`data` is either `Paperclip::UploadedFileAdapter` or `Paperclip::FileAdapter`. That does not play well with this bit of code: https://github.com/google/google-api-ruby-client/blob/master/lib/google/apis/core/upload.rb#L52-L64 as it raises `Invalid upload source` Using the path to the file seems to be working fine, but I'm not entirely sure if this is the best way to fix this.

icco commented:
Hmm. So it looks like some of our tests expect different things. Could you make this a check similar to the google api client? like if streamable, pass data, if path exists, pass path? or something? And then also write a test?

```
 30) Error:
TestStorageRequests#test_files_set_body_file:
NoMethodError: undefined method `path' for #<StringIO:0x00000002f1c708>
    /tmp/build/453021c2/src/fog-google/lib/fog/storage/google_json/requests/put_object.rb:35:in `put_object'
    /tmp/build/453021c2/src/fog-google/lib/fog/storage/google_json/models/file.rb:70:in `save'
    /tmp/build/453021c2/src/fog-google/test/integration/storage/test_files.rb:82:in `test_files_set_body_file'

 31) Error:
TestStorageRequests#test_files_create_string:
NoMethodError: undefined method `path' for #<StringIO:0x00000006e487d8>
    /tmp/build/453021c2/src/fog-google/lib/fog/storage/google_json/requests/put_object.rb:35:in `put_object'
    /tmp/build/453021c2/src/fog-google/lib/fog/storage/google_json/models/file.rb:70:in `save'
    /var/lib/gems/2.3.0/gems/fog-core-1.44.3/lib/fog/core/collection.rb:50:in `create'
    /tmp/build/453021c2/src/fog-google/test/integration/storage/test_files.rb:22:in `test_files_create_string'

 32) Error:
TestStorageRequests#test_files_set_body_string:
NoMethodError: undefined method `path' for #<StringIO:0x00000006f16228>
    /tmp/build/453021c2/src/fog-google/lib/fog/storage/google_json/requests/put_object.rb:35:in `put_object'
    /tmp/build/453021c2/src/fog-google/lib/fog/storage/google_json/models/file.rb:70:in `save'
    /var/lib/gems/2.3.0/gems/fog-core-1.44.3/lib/fog/core/collection.rb:50:in `create'
    /tmp/build/453021c2/src/fog-google/test/integration/storage/test_files.rb:57:in `test_files_set_body_string'

 33) Error:
TestStorageRequests#test_put_object_file:
NoMethodError: undefined method `path' for #<StringIO:0x00000009592ca8>
    /tmp/build/453021c2/src/fog-google/lib/fog/storage/google_json/requests/put_object.rb:35:in `put_object'
    /tmp/build/453021c2/src/fog-google/test/integration/storage/test_objects.rb:21:in `test_put_object_file'
```

icco commented:
(all of our CIs are semi-broken, sorry about that)

icco commented:
@Everlag yeah or just the block that @DawidJanczak links to in the google api client. Either would work.

DawidJanczak commented:
Sorry about that, for some reason I assumed this is always in paperclip world. I've changed the implementation and added a test, PTAL.

Mind that I didn't run the tests locally - is there any way of running them without valid credentials? I don't want to use my existing project as I have no way of knowing what the tests actually change in buckets (if at all).

icco commented:
Historically we have had mocks to solve that problem, but with this big rewrite, we haven't come up with a good mock solution. Personally, I just create a new Google cloud project and use those creds to test.

icco commented on test/integration/storage/test_objects.rb:
```diff
@@ -25,6 +25,18 @@ def test_put_object_file
+    object_name = new_object_name
+    paperclip_file = OpenStruct.new(:path => "/tmp/fog-google-storage",
+                                    :content_type => "image/png")
+    @cliient.put_object(some_bucket_name, object_name, paperclip_file)
```
Typo on client.

DawidJanczak commented:
All fixed @icco And thanks for the explanation, I'll do the same thing to set tests up.
